go/store/types/ordered_sequences.go

```diff
@@ -89,6 +89,72 @@ func newCursorAt(ctx context.Context, seq orderedSequence, key orderedKey, forIn
 	return cur, nil
 }
 
+// newCursorBackFromValue returns a reverse sequence iterator starting from the greatest value <= the one given
+func newCursorBackFromValue(ctx context.Context, seq orderedSequence, val Value) (*sequenceCursor, error) {
+	var key orderedKey
+	if val != nil {
+		var err error
+		key, err = newOrderedKey(val, seq.format())
+
+		if err != nil {
+			return nil, err
+		}
+	}
+	return newCursorBackFrom(ctx, seq, key)
+}
+
+func newCursorBackFrom(ctx context.Context, seq orderedSequence, key orderedKey) (*sequenceCursor, error) {
+	var cur *sequenceCursor
+	for {
+		idx := 0
+		cur = newReverseSequenceCursor(cur, seq, idx)
+		if key != emptyKey {
+			// If we run off the end of the sequence, start the cursor at the last element.
+			ok, err := seekTo(cur, key, true)
+
+			if err != nil {
+				return nil, err
+			}
+
+			if !ok {
+				return cur, nil
+			}
+		}
+
+		cs, err := cur.getChildSequence(ctx)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if cs == nil {
+			break
+		}
+		seq = cs.(orderedSequence)
+	}
+
+	// If we overshot the key, back off by one. We want the greatest element <= key
+	if cur.valid() {
+		currKey, err := cur.seq.(orderedSequence).getKey(cur.idx)
+		d.PanicIfError(err)
```

Comment:
Is this an appropriate panic, or should be return err?

Reply:
Good call